simulator/syncer/syncer_test.go

```diff
@@ -0,0 +1,481 @@
+							{
+								Name: "container-1",
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			s := runtime.NewScheme()
+			v1.AddToScheme(s)
+			scheduling.AddToScheme(s)
+			storage.AddToScheme(s)
+			src := dynamicFake.NewSimpleDynamicClient(s)
+			dest := dynamicFake.NewSimpleDynamicClient(s)
+			resources := []*restmapper.APIGroupResources{
+				{
+					Group: metav1.APIGroup{
+						Versions: []metav1.GroupVersionForDiscovery{
+							{Version: "v1"},
+						},
+					},
+					VersionedResources: map[string][]metav1.APIResource{
+						"v1": {
+							{Name: "pods", Namespaced: true, Kind: "Pod"},
+						},
+					},
+				},
+				{
+					Group: metav1.APIGroup{
+						Versions: []metav1.GroupVersionForDiscovery{
+							{Version: "v1"},
+						},
+					},
+					VersionedResources: map[string][]metav1.APIResource{
+						"v1": {
+							{Name: "nodes", Namespaced: true, Kind: "Node"},
+						},
+					},
+				},
+			}
+			mapper := restmapper.NewDiscoveryRESTMapper(resources)
+			service := New(src, dest, mapper)
+
+			ctx, cancel := context.WithCancel(context.Background())
+
+			createdPods := sets.New[podKey]()
+			for _, pod := range tt.initialPodsInSrcCluster {
+				p, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
+				if err != nil {
+					t.Fatalf("failed to convert pod to unstructured: %v", err)
+				}
+				unstructedPod := &unstructured.Unstructured{Object: p}
+				_, err = src.Resource(v1.Resource("pods").WithVersion("v1")).Namespace(pod.Namespace).Create(ctx, unstructedPod, metav1.CreateOptions{})
+				if err != nil {
+					t.Fatalf("failed to create pod: %v", err)
+				}
+				createdPods.Insert(podKey{pod.Name, pod.Namespace})
+			}
+
+			go service.Run(ctx)
+			defer cancel()
+
+			for _, pod := range tt.podsCreatedInSrcCluster {
+				p, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
+				if err != nil {
+					t.Fatalf("failed to convert pod to unstructured: %v", err)
+				}
+				unstructedPod := &unstructured.Unstructured{Object: p}
+				_, err = src.Resource(v1.Resource("pods").WithVersion("v1")).Namespace(pod.Namespace).Create(ctx, unstructedPod, metav1.CreateOptions{})
+				if err != nil {
+					t.Fatalf("failed to create pod: %v", err)
+				}
+				createdPods.Insert(podKey{pod.Name, pod.Namespace})
+			}
+
+			time.Sleep(500 * time.Millisecond)
+
+			for _, pod := range tt.podsUpdatedInSrcCluster {
+				p, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
+				if err != nil {
+					t.Fatalf("failed to convert pod to unstructured: %v", err)
+				}
+				unstructedPod := &unstructured.Unstructured{Object: p}
+				_, err = src.Resource(v1.Resource("pods").WithVersion("v1")).Namespace(pod.Namespace).Update(ctx, unstructedPod, metav1.UpdateOptions{})
+				if err != nil {
+					t.Fatalf("failed to update pod: %v", err)
+				}
+			}
+			time.Sleep(500 * time.Millisecond)
```

Comment:
I'd like to avoid `time.Sleep` because the time needed depends on the environment in which this test runs. Do you have any ideas to avoid it?

Reply:
`time.Sleep` between creating and updating, and one between updating and deleting might be unnecessary.

The problem is how to wait for the syncer to apply source cluster's resources to the destination cluster. This could be done by trying some times using `PollUntilContextTimeout`.